File: solutions/templates/SOLUTION_TEMPLATE.py
"""
# Difficulty: [Easy/Medium/Hard]

[Problem description goes here. Use markdown formatting with `code` for inline code.]

[Additional problem constraints or notes.]

**Example:**

<dl class="example-details">
<dt>Input:</dt>
<dd>[input example]</dd>
<dt>Output:</dt>
<dd>result1 = solution.methodName(param1_1, param2_1)</dd>
<dt>Explanation:</dt>
<dd>[explanation of how input becomes output]</dd>
</dl>

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
[High-level insight or key observation that makes the solution work.
Keep this section brief - 1-3 sentences maximum.
Explain the "aha moment" or central idea.]

### APPROACH:
[Detailed explanation of the solution approach written in flowing prose.
Break down the algorithm step by step.
Explain why each step is necessary.
Use clear, descriptive language without numbered lists.
Focus on helping the reader understand the thought process.]

### WHY THIS WORKS:
[Brief explanation of correctness.
Why does this approach solve the problem?
What properties make it work?
Use bullet points for clarity.]

### EXAMPLE WALKTHROUGH:

Input:
```
[input example]
```

**Step 1:** [description of first step]

**Step 2:** [description of second step]

**Step N:** [description of final step]

### TIME COMPLEXITY:
**O(?)** - [explanation]

### SPACE COMPLEXITY:
**O(?)** - [explanation]

### EDGE CASES:
- **[Edge case 1]:** [how it's handled]
- **[Edge case 2]:** [how it's handled]
- **[Edge case 3]:** [how it's handled]

</details>
"""

from typing import Any
import re



class Solution:
    def methodName(self, param1: Any, param2: Any) -> Any:
        """
        Approach: [Brief description of approach]
        Time Complexity: O(?)
        Space Complexity: O(?)

        Args:
            param1: [description and type]
            param2: [description and type]

        Returns:
            [description and type]
        """
        # Implementation with clear comments
        pass

    def methodNameAlternative(self, param1: Any, param2: Any) -> Any:
        """
        [Alternative Approach Name]
        Time Complexity: O(?)
        Space Complexity: O(?)

        Args:
            param1: [description and type]
            param2: [description and type]

        Returns:
            [description and type]
        """
        # Alternative implementation
        pass


# Test cases
if __name__ == "__main__":
    solution = Solution()

    # Test case 1
    param1_1 = None  # Replace with actual test input
    param2_1 = None  # Replace with actual test input
    expected1 = None  # Replace with expected output
    result1 = solution.methodName(param1_1, param2_1)
    print(f"Input: param1={param1_1}, param2={param2_1}")
    print(f"Output: {result1}")
    print(f"Expected: {expected1}")
    print()

    # Test case 2
    param1_2 = None
    param2_2 = None
    expected2 = None
    result2 = solution.methodName(param1_2, param2_2)
    print(f"Input: param1={param1_2}, param2={param2_2}")
    print(f"Output: {result2}")
    print(f"Expected: {expected2}")
    print()

    # Test case 3 - Edge case
    param1_3 = None
    param2_3 = None
    expected3 = None
    result3 = solution.methodName(param1_3, param2_3)
    print(f"Input: param1={param1_3}, param2={param2_3}")
    print(f"Output: {result3}")
    print(f"Expected: {expected3}")

File: docs/developer-guide/templates/SOLUTION_TEMPLATE.java
/**
 * Difficulty: [Easy/Medium/Hard]
 *
 * [Problem description goes here. Use markdown formatting with `code` for inline code.]
 *
 * [Additional problem constraints or notes.]
 *
 * **Example:**
 *
 * <dl class="example-details">
 * <dt>Input:</dt>
 * <dd>[input example]</dd>
 * <dt>Output:</dt>
 * <dd>[output example]</dd>
 * <dt>Explanation:</dt>
 * <dd>[explanation of how input becomes output]</dd>
 * </dl>
 *
 * <details>
 * <summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### METADATA:
**Techniques**: [e.g., Hash Table Lookup, Binary Search, Two Pointers]
**Data Structures**: [e.g., Hash Map, Array, Tree, Graph]
**Patterns**: [e.g., Complement Search, Sliding Window, DFS]
**Time Complexity**: **O(?)** - [detailed explanation of time complexity]
**Space Complexity**: **O(?)** - [detailed explanation of space complexity]

 *
 * ### INTUITION:
 * [High-level insight or key observation that makes the solution work.]
 *
 * [Keep this section brief - 1-3 sentences maximum.]
 *
 * [Explain the "aha moment" or central idea.]
 *
 * ### APPROACH:
 * [Detailed explanation of the solution approach written in flowing prose.]
 *
 * [Break down the algorithm step by step.]
 *
 * [Explain why each step is necessary.]
 *
 * [Use clear, descriptive language without numbered lists.]
 *
 * [Focus on helping the reader understand the thought process.]
 *
 * ### WHY THIS WORKS:
 * [Brief explanation of correctness - use bullet points for clarity.]
 *
 * [Why does this approach solve the problem?]
 *
 * [What properties make it work?]
 *
 * ### EXAMPLE WALKTHROUGH:
 * Input:
 * ```
 * [input example]
 * ```
 *
 * **Step 1:** [description of first step]
 *
 * **Step 2:** [description of second step]
 *
 * **Step N:** [description of final step]
 *
 * ### TIME COMPLEXITY:
 * **O(?)** - [detailed explanation]
 *
 * ### SPACE COMPLEXITY:
 * **O(?)** - [detailed explanation]
 *
 * ### EDGE CASES:
 * - **[Edge case 1]:** [how it's handled]
 * - **[Edge case 2]:** [how it's handled]
 * - **[Edge case 3]:** [how it's handled]
 *
 * </details>
 */

import java.util.*;

class Solution {
    /**
     * Main solution method
     *
     * Approach: [Brief description of approach]
     * Time Complexity: O(?)
     * Space Complexity: O(?)
     *
     * @param param1 [description and type]
     * @param param2 [description and type]
     * @return [description and type]
     */
    public ReturnType methodName(ParamType1 param1, ParamType2 param2) {
        // Implementation with clear comments
        return null;
    }

    /**
     * Alternative solution method
     *
     * Approach: [Alternative Approach Name]
     * Time Complexity: O(?)
     * Space Complexity: O(?)
     *
     * @param param1 [description and type]
     * @param param2 [description and type]
     * @return [description and type]
     */
    public ReturnType methodNameAlternative(ParamType1 param1, ParamType2 param2) {
        // Alternative implementation
        return null;
    }

    // Test cases
    public static void main(String[] args) {
        Solution solution = new Solution();

        // Test case 1
        System.out.println("Test Case 1:");
        ParamType1 param1_1 = null; // Replace with actual test input
        ParamType2 param2_1 = null; // Replace with actual test input
        ReturnType expected1 = null; // Replace with expected output
        ReturnType result1 = solution.methodName(param1_1, param2_1);
        System.out.println("Input: param1=" + param1_1 + ", param2=" + param2_1);
        System.out.println("Output: " + result1);
        System.out.println("Expected: " + expected1);
        System.out.println("Pass: " + (Objects.equals(result1, expected1)));
        System.out.println();

        // Test case 2
        System.out.println("Test Case 2:");
        ParamType1 param1_2 = null;
        ParamType2 param2_2 = null;
        ReturnType expected2 = null;
        ReturnType result2 = solution.methodName(param1_2, param2_2);
        System.out.println("Input: param1=" + param1_2 + ", param2=" + param2_2);
        System.out.println("Output: " + result2);
        System.out.println("Expected: " + expected2);
        System.out.println("Pass: " + (Objects.equals(result2, expected2)));
        System.out.println();

        // Test case 3 - Edge case
        System.out.println("Test Case 3 (Edge Case):");
        ParamType1 param1_3 = null;
        ParamType2 param2_3 = null;
        ReturnType expected3 = null;
        ReturnType result3 = solution.methodName(param1_3, param2_3);
        System.out.println("Input: param1=" + param1_3 + ", param2=" + param2_3);
        System.out.println("Output: " + result3);
        System.out.println("Expected: " + expected3);
        System.out.println("Pass: " + (Objects.equals(result3, expected3)));
    }
}
